pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -922,19 +929,94 @@ func (l *LibvirtDomainManager) SyncVMI(vmi *v1.VirtualMachineInstance, allowEmul
 			return nil, err
 		}
-		err = dom.DetachDeviceFlags(strings.ToLower(string(detachBytes)), affectLiveAndConfigLibvirtFlags)
-		if err != nil {
-			logger.Reason(err).Error("detaching device")
-			return nil, err
-		}
+
+		// Create goroutine to do detachment so we can wait to see if the disk is reattached.
+		go func(namespacedDiskName, detachDiskName, detachDiskTargetDevice, domainName string, detachBytes []byte) {
+			logger.V(hotplugDetachmentVerbosity).Infof("Potentially detaching hotplug disk %s", namespacedDiskName)
+			// Claim mutex again to work on protected data like the channels map and the domain.
+			l.domainModifyLock.Lock()
+			// If channel has already been created, a goroutine is already waiting for the disk to be detached.  Exit
+			// so there aren't multiple goroutines waiting to detach the same disk.
+			if _, ok := l.cancelDiskDetachChannels[namespacedDiskName]; ok {
+				logger.V(hotplugDetachmentVerbosity).Infof("Potential hotplug disk detaching already handled: %s", namespacedDiskName)
+				// Unlock the mutex so protected data can be accessed
+				l.domainModifyLock.Unlock()
+				return
+			}
+			// There are no other goroutines waiting to detach this disk, create a channel enable subsequent calls to
+			// SyncVMI to cancel this detach (if attach is requested within the grace period: 5 seconds)
+			cancelChannel := make(chan struct{})
+			l.cancelDiskDetachChannels[namespacedDiskName] = cancelChannel
+			// Unlock the mutex during the grace period wait so SyncVMI can continue to run
+			l.domainModifyLock.Unlock()
+
+			doDetach := false
+			// Wait for either the grace period to expire (in which case: proceed to detach) or for the attach channel
+			// to be triggered (in which case: cancel the detach)
+			select {
+			case <-cancelChannel:
+				// Disk attach has been requested within the grace period, cancel the detach
+				logger.V(hotplugDetachmentVerbosity).Infof("Potential hotplug disk detaching was cancelled: %s", namespacedDiskName)
+			case <-time.After(hotplugDetachmentGracePeriod):
+				// Grace period has expired without an attach request, proceed to detach
+				doDetach = true
+			}
+
+			// Claim mutex again to work on protected data like the channels map and the domain.
+			l.domainModifyLock.Lock()
+			// Ensure that the detach cancel channel is cleared and the mutex is unlocked before exiting
+			defer func() {
+				// Detach request has been handled, remove channel from map
+				delete(l.cancelDiskDetachChannels, namespacedDiskName)
+				// Unlock the mutex so protected data can be accessed
+				l.domainModifyLock.Unlock()
+			}()
+
+			if doDetach {
+				// Detach the volume
+				logger.V(1).Infof("Detaching disk %s, target %s", detachDiskName, detachDiskTargetDevice)
+				dom, err := l.virConn.LookupDomainByName(domainName)
+				if err != nil {
+					logger.Reason(err).Error("getting domain for device detach")
+					return
+				}
+				defer dom.Free()
```

Comment:
does `defer dom.Free()` race with the defer on line 998?  maybe move the dom.Free call back down to line 1018 and ditch this defer call to ensure it is protected by the mutex.